File: backend/app/api/routes/policy.py
from __future__ import annotations

from typing import Literal

from fastapi import APIRouter, Depends, HTTPException, Query, status
from sqlalchemy import func, select
from sqlalchemy.ext.asyncio import AsyncSession

from app.db.session import get_db_session
from app.models.policy_document import PolicyDocument
from app.models.policy_document_attachment import PolicyDocumentAttachment
from app.schemas.policy import (
    PolicyDocumentAttachmentResponse,
    PolicyDocumentDetailResponse,
    PolicyDocumentListItemResponse,
    PolicyDocumentPageResponse,
    PolicyFilterResponse,
    PolicyFilterValueResponse,
)


router = APIRouter(prefix="/policy", tags=["policy"])


def _serialize_policy_document_item(
    document: PolicyDocument,
) -> PolicyDocumentListItemResponse:
    return PolicyDocumentListItemResponse(
        id=document.id,
        source=document.source,
        title=document.title,
        summary=document.summary,
        document_no=document.document_no,
        issuing_authority=document.issuing_authority,
        policy_level=document.policy_level,
        category=document.category,
        macro_topic=document.macro_topic,
        published_at=document.published_at,
        effective_at=document.effective_at,
        url=document.url,
        metadata_status=document.metadata_status,
        projection_status=document.projection_status,
    )


@router.get("/documents", response_model=PolicyDocumentPageResponse)
async def list_policy_documents(
    authority: str | None = Query(default=None),
    category: str | None = Query(default=None),
    macro_topic: str | None = Query(default=None),
    search_scope: Literal["basic", "fulltext"] = Query(default="basic"),
    published_from: str | None = Query(default=None),
    published_to: str | None = Query(default=None),
    keyword: str | None = Query(default=None),
    page: int = Query(default=1, ge=1),
    page_size: int = Query(default=20, ge=1, le=100),
    session: AsyncSession = Depends(get_db_session),
) -> PolicyDocumentPageResponse:
    filters: list[object] = []

    normalized_authority = authority.strip() if authority else ""
    if normalized_authority:
        filters.append(PolicyDocument.issuing_authority == normalized_authority)

    normalized_category = category.strip() if category else ""
    if normalized_category:
        filters.append(PolicyDocument.category == normalized_category)

    normalized_macro_topic = macro_topic.strip() if macro_topic else ""
    if normalized_macro_topic:
        filters.append(PolicyDocument.macro_topic == normalized_macro_topic)

    if keyword and keyword.strip():
        wildcard = f"%{keyword.strip()}%"
        keyword_statement = (
            PolicyDocument.title.ilike(wildcard)
            | PolicyDocument.summary.ilike(wildcard)
            | PolicyDocument.document_no.ilike(wildcard)
        )
        # 关键流程：默认仅走轻量字段匹配，只有显式开启全文检索时才扫描正文，降低列表首屏等待时间。
        if search_scope == "fulltext":
            keyword_statement = keyword_statement | PolicyDocument.content_text.ilike(
                wildcard
            )
        filters.append(keyword_statement)

    if published_from:
        filters.append(PolicyDocument.published_at >= published_from)
    if published_to:
        filters.append(PolicyDocument.published_at <= published_to)

    statement = select(PolicyDocument)
    count_statement = select(func.count(PolicyDocument.id))
    if filters:
        statement = statement.where(*filters)
        count_statement = count_statement.where(*filters)

    # 关键流程：总数统计改为直接 count 主键，避免为分页额外包一层子查询。
    total = int((await session.execute(count_statement)).scalar_one())
    rows = (
        await session.execute(
            statement
            .order_by(PolicyDocument.published_at.desc(), PolicyDocument.created_at.desc())
            .offset((page - 1) * page_size)
            .limit(page_size)
        )
    ).scalars().all()

    return PolicyDocumentPageResponse(
        items=[_serialize_policy_document_item(item) for item in rows],
        total=total,
        page=page,
        page_size=page_size,
    )


@router.get("/documents/{document_id}", response_model=PolicyDocumentDetailResponse)
async def get_policy_document_detail(
    document_id: str,
    session: AsyncSession = Depends(get_db_session),
) -> PolicyDocumentDetailResponse:
    document = await session.get(PolicyDocument, document_id)
    if document is None:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="policy document not found",
        )

    attachments = (
        await session.execute(
            select(PolicyDocumentAttachment)
            .where(PolicyDocumentAttachment.document_id == document_id)
            .order_by(PolicyDocumentAttachment.created_at.asc())
        )
    ).scalars().all()
    return PolicyDocumentDetailResponse(
        **_serialize_policy_document_item(document).model_dump(),
        content_text=document.content_text,
        content_html=document.content_html,
        attachments=[
            PolicyDocumentAttachmentResponse(
                attachment_url=item.attachment_url,
                attachment_name=item.attachment_name,
                attachment_type=item.attachment_type,
            )
            for item in attachments
        ],
        industry_tags=list(document.industry_tags_json or []),
        market_tags=list(document.market_tags_json or []),
    )


@router.get("/filters", response_model=PolicyFilterResponse)
async def get_policy_filters(
    session: AsyncSession = Depends(get_db_session),
) -> PolicyFilterResponse:
    async def load_distinct_values(column) -> list[str]:
        rows = (
            await session.execute(
                select(column)
                .where(column.is_not(None))
                .distinct()
                .order_by(column.asc())
            )
        ).scalars().all()
        return [str(item) for item in rows if str(item).strip()]

    authorities = await load_distinct_values(PolicyDocument.issuing_authority)
    categories = await load_distinct_values(PolicyDocument.category)
    macro_topics = await load_distinct_values(PolicyDocument.macro_topic)

    return PolicyFilterResponse(
        authorities=[
            PolicyFilterValueResponse(label=item, value=item) for item in authorities
        ],
        categories=[
            PolicyFilterValueResponse(label=item, value=item) for item in categories
        ],
        macro_topics=[
            PolicyFilterValueResponse(label=item, value=item) for item in macro_topics
        ],
    )

File: frontend/src/api/policy.ts
import { requestJson } from './http'
import { buildQueryString } from './query'

export type PolicyDocumentListItem = {
  id: string
  source: string
  title: string
  summary: string | null
  document_no: string | null
  issuing_authority: string | null
  policy_level: string | null
  category: string | null
  macro_topic: string | null
  published_at: string | null
  effective_at: string | null
  url: string
  metadata_status: string
  projection_status: string
}

export type PolicyDocumentAttachment = {
  attachment_url: string
  attachment_name: string | null
  attachment_type: string | null
}

export type PolicyDocumentDetail = PolicyDocumentListItem & {
  content_text: string | null
  content_html: string | null
  attachments: PolicyDocumentAttachment[]
  industry_tags: string[]
  market_tags: string[]
}

export type PolicyFilterOption = {
  label: string
  value: string
}

export type PolicyFilters = {
  authorities: PolicyFilterOption[]
  categories: PolicyFilterOption[]
  macro_topics: PolicyFilterOption[]
}

export type PolicyDocumentPage = {
  items: PolicyDocumentListItem[]
  total: number
  page: number
  page_size: number
}

export type PolicyQuery = {
  authority?: string
  category?: string
  macroTopic?: string
  keyword?: string
  searchScope?: 'basic' | 'fulltext'
  page?: number
  pageSize?: number
}

export type PolicySyncResponse = {
  job_id: string | null
  job_type: string
  status: string
  provider_count: number
  raw_count: number
  normalized_count: number
  inserted_count: number
  updated_count: number
  deduped_count: number
  failed_provider_count: number
}

const isRecord = (value: unknown): value is Record<string, unknown> =>
  typeof value === 'object' && value !== null && !Array.isArray(value)

const asString = (value: unknown, fallback = '') =>
  typeof value === 'string' ? value : fallback

const asNullableString = (value: unknown) =>
  typeof value === 'string' ? value : null

const asNumber = (value: unknown, fallback = 0) =>
  typeof value === 'number' && Number.isFinite(value) ? value : fallback

const asStringArray = (value: unknown) =>
  Array.isArray(value) ? value.filter((item): item is string => typeof item === 'string') : []

const normalizePolicyDocumentListItem = (item: unknown): PolicyDocumentListItem => {
  if (!isRecord(item)) {
    return {
      id: '',
      source: '',
      title: '',
      summary: null,
      document_no: null,
      issuing_authority: null,
      policy_level: null,
      category: null,
      macro_topic: null,
      published_at: null,
      effective_at: null,
      url: '',
      metadata_status: '',
      projection_status: '',
    }
  }

  return {
    id: asString(item.id),
    source: asString(item.source),
    title: asString(item.title),
    summary: asNullableString(item.summary),
    document_no: asNullableString(item.document_no),
    issuing_authority: asNullableString(item.issuing_authority),
    policy_level: asNullableString(item.policy_level),
    category: asNullableString(item.category),
    macro_topic: asNullableString(item.macro_topic),
    published_at: asNullableString(item.published_at),
    effective_at: asNullableString(item.effective_at),
    url: asString(item.url),
    metadata_status: asString(item.metadata_status),
    projection_status: asString(item.projection_status),
  }
}

const normalizePolicyFilterOption = (item: unknown): PolicyFilterOption => {
  if (!isRecord(item)) {
    return { label: '', value: '' }
  }
  return {
    label: asString(item.label),
    value: asString(item.value),
  }
}

const normalizePolicyDocumentDetail = (item: unknown): PolicyDocumentDetail => {
  const base = normalizePolicyDocumentListItem(item)
  if (!isRecord(item)) {
    return {
      ...base,
      content_text: null,
      content_html: null,
      attachments: [],
      industry_tags: [],
      market_tags: [],
    }
  }

  return {
    ...base,
    content_text: asNullableString(item.content_text),
    content_html: asNullableString(item.content_html),
    attachments: Array.isArray(item.attachments)
      ? item.attachments
          .filter((attachment): attachment is Record<string, unknown> => isRecord(attachment))
          .map((attachment) => ({
            attachment_url: asString(attachment.attachment_url),
            attachment_name: asNullableString(attachment.attachment_name),
            attachment_type: asNullableString(attachment.attachment_type),
          }))
      : [],
    industry_tags: asStringArray(item.industry_tags),
    market_tags: asStringArray(item.market_tags),
  }
}

export const policyApi = {
  async getDocuments(query: PolicyQuery): Promise<PolicyDocumentPage> {
    // 关键流程：在 API 层统一拼接分页和筛选参数，避免视图层重复维护接口兼容细节。
    const search = buildQueryString({
      authority: query.authority || undefined,
      category: query.category || undefined,
      macro_topic: query.macroTopic || undefined,
      keyword: query.keyword || undefined,
      search_scope: query.searchScope ?? 'basic',
      page: query.page ?? 1,
      page_size: query.pageSize ?? 12,
    })
    const payload = await requestJson<unknown>(`/api/policy/documents${search}`)
    if (!isRecord(payload)) {
      return { items: [], total: 0, page: 1, page_size: 12 }
    }
    return {
      items: Array.isArray(payload.items)
        ? payload.items.map(normalizePolicyDocumentListItem)
        : [],
      total: asNumber(payload.total),
      page: asNumber(payload.page, 1),
      page_size: asNumber(payload.page_size, 12),
    }
  },

  async getDocument(documentId: string): Promise<PolicyDocumentDetail> {
    const payload = await requestJson<unknown>(
      `/api/policy/documents/${encodeURIComponent(documentId)}`,
    )
    return normalizePolicyDocumentDetail(payload)
  },

  async getFilters(): Promise<PolicyFilters> {
    const payload = await requestJson<unknown>('/api/policy/filters')
    if (!isRecord(payload)) {
      return { authorities: [], categories: [], macro_topics: [] }
    }
    return {
      authorities: Array.isArray(payload.authorities)
        ? payload.authorities.map(normalizePolicyFilterOption)
        : [],
      categories: Array.isArray(payload.categories)
        ? payload.categories.map(normalizePolicyFilterOption)
        : [],
      macro_topics: Array.isArray(payload.macro_topics)
        ? payload.macro_topics.map(normalizePolicyFilterOption)
        : [],
    }
  },

  async syncDocuments(
    forceRefresh = false,
    accessToken?: string | null,
  ): Promise<PolicySyncResponse> {
    const payload = await requestJson<unknown>('/api/admin/policy/sync', {
      method: 'POST',
      body: { force_refresh: forceRefresh },
      accessToken: accessToken ?? undefined,
    })
    if (!isRecord(payload)) {
      return {
        job_id: null,
        job_type: 'policy_sync',
        status: 'failed',
        provider_count: 0,
        raw_count: 0,
        normalized_count: 0,
        inserted_count: 0,
        updated_count: 0,
        deduped_count: 0,
        failed_provider_count: 0,
      }
    }
    return {
      job_id: asNullableString(payload.job_id),
      job_type: asString(payload.job_type, 'policy_sync'),
      status: asString(payload.status, 'failed'),
      provider_count: asNumber(payload.provider_count),
      raw_count: asNumber(payload.raw_count),
      normalized_count: asNumber(payload.normalized_count),
      inserted_count: asNumber(payload.inserted_count),
      updated_count: asNumber(payload.updated_count),
      deduped_count: asNumber(payload.deduped_count),
      failed_provider_count: asNumber(payload.failed_provider_count),
    }
  },
}
